codeU1.py

```diff
@@ -76,43 +78,44 @@ def insert(self, value):
         new_node = Node(value)
         if not self.head:
             self.head = new_node;
+            self.tail = self.head;
             self.N = self.N+1;
-            return;
-        cur = self.head
-        while cur.next:
-            cur = cur.next   
-        cur.next = new_node 
-        self.N = self.N+1;
+            return;  
+        self.tail.next = new_node
+        self.tail = new_node
+        self.N+=1;
         
     def print_list(self):
         if not self.head:
-            print 'Empty list'
+            logging.warning('Empty list')
         cur = self.head
         while cur:
             print cur.val,
             cur = cur.next 
         print '\n'
-        return;    
+        return;  
+    
         
-#def k_last_element(x,k):
-#    print str(k) + 'th last element is ' + str(ll.k_last(k))     
-
-def test_list():      
-   x = [0,1,2,3]
-   ll = LinkedList();
-   for a in x:
-       ll.insert(a)
-   print 'List:'
-   ll.print_list()
-
-   print '0th last element is ' + str(ll.get_kth_last_element(0))
-   print '1th last element is ' + str(ll.get_kth_last_element(1))
-   print '2th last element is ' + str(ll.get_kth_last_element(4))
-
-
-print "="*25 + '\n Test singly-linked list \n' + "="*25
-test_list()
+class TestList(unittest.TestCase):
+    x = [0,1,2,3]

```

Comment:
`unittest.TestCase` defines a `setUp` method that will be called before each test case:

https://docs.python.org/2/library/unittest.html#unittest.TestCase.setUp

This is frequently used for common initialization that is required before each test case, like so:

```
def setUp(self):
  self.ll = LinkedList()
  for a in range(4):
    ll.insert(a)
```

With this initialization, the test cases below can just refer to `self.ll`.